google/cloud/stream_range.h

```diff
@@ -186,6 +186,7 @@ class StreamRange {
     };
     auto v = reader_();
     absl::visit(UnpackVariant{*this}, std::move(v));
+    current_ok_ = current_.ok();
```

Comment:
Good observation. I can't unconditionally assign to `sr.current_`, because it's not allowed to assign an OK Status to a StatusOr. However, I did change the code a bit to be clearer and more correct by moving the `sr.current_ok_` assignment into the `UnpackVariant` functor, so it can set it correctly in each case. PTAL.

Reply:
OK, sounds good.

Aside: In the bigger picture I still find the control flow confusing.  For example, there are really only three states ...
```
    { => T } S1 [ => non-OK ] S2 [ => end ] S3
```
... but we currently use all four.  To that end, I might make the initial state S1 ...
```
  bool current_ok_ = true;
  bool is_end_ = false;
```
and also stop S3 from ever calling `reader_()`.